File: sdk/golang/destination.go
// MIT License
//
// Copyright (c) 2021 Shyam Jesalpura and EASE lab
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

package golang

import (
	"capnproto.org/go/capnp/v3/rpc"
	"context"
	"github.com/ease-lab/vhive-xdt/proto/crossXDT"
	"google.golang.org/grpc/metadata"
	"net"
	"strings"

	"github.com/ease-lab/vhive-xdt/proto/downXDT"
	"github.com/ease-lab/vhive-xdt/utils"

	log "github.com/sirupsen/logrus"
	"go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc"

	"google.golang.org/grpc"
)

type downXDTServer struct {
	config  utils.Config
	handler func([]byte) ([]byte, bool)
	conn    *rpc.Conn
	downXDT.UnimplementedXDTtoFnServer
}

// XDTFnCall is to be called by dQP to invoke DstFn
func (s downXDTServer) XDTFnCall(ctx context.Context, in *downXDT.InvocationRequest) (*downXDT.InvocationResponse, error) {

	log.Infof("DST: received invocation call %s", in.XDTJSON)

	headers, ok := metadata.FromIncomingContext(ctx)
	var message []byte
	if ok && headers["is_xdt"][0] == "true" {
		key := headers["key"][0]
		log.Infof("DST: using %s routing", headers["routing"][0])

		// fetch data from dQP
		payloadBytes, err := FetchData(ctx, key, s.conn)
		if err != nil {
			log.Errorf("DST: FetchData failed %v", err)
			return &downXDT.InvocationResponse{}, err
		}

		//call destination function
		message, ok = s.handler(payloadBytes)
	}

	return &downXDT.InvocationResponse{Message: message, Ok: ok}, nil
}

// FetchData fetches data from dQP to DstFn
func FetchData(ctx context.Context, key string, capconn *rpc.Conn) ([]byte, error) {
	client := crossXDT.StreamData(capconn.Bootstrap(context.Background()))
	f, _ := client.ServeData(ctx, func(ps crossXDT.StreamData_serveData_Params) error {
		ps.SetKey(key)
		return nil
	})
	//defer release()

	res, err := f.Struct()
	if err != nil {
		log.Fatal(err)
	}
	payload, err := res.Payload()
	if err != nil {
		log.Fatal(err)
	}

	log.Info("DST: ", payload[0:9], payload[len(payload)-9:])
	return payload, nil
}

// Get pulls payload from DQP server using the key
func Get(ctx context.Context, capability string, config utils.Config) ([]byte, error) {
	log.Infof("attempting Get using capability %s", capability)
	key := capability
	splitString := strings.SplitN(capability, "|", 2)
	sQPAddr := splitString[1]
	httpMetadata := map[string]string{
		"is_xdt":   "true",
		"key":      key,
		"sqp_addr": sQPAddr,
		"routing":  config.Routing,
	}
	ctx = metadata.NewOutgoingContext(ctx, metadata.New(httpMetadata))

	conn, err := net.Dial("tcp", sQPAddr)
	if err != nil {
		log.Fatal(err)
	}
	capconn := rpc.NewConn(rpc.NewStreamTransport(conn), nil)
	return FetchData(ctx, key, capconn)
}

// BroadcastGet pulls payload from DQP server using the key
func BroadcastGet(ctx context.Context, capability string, config utils.Config) ([]byte, error) {
	log.Infof("attempting Get using capability %s", capability)
	key := capability
	splitString := strings.SplitN(capability, "|", 2)
	sQPAddr := splitString[1]
	httpMetadata := map[string]string{
		"is_xdt":   "true",
		"key":      key,
		"sqp_addr": sQPAddr,
		"routing":  config.Routing,
	}
	ctx = metadata.NewOutgoingContext(ctx, metadata.New(httpMetadata))

	conn, err := net.Dial("tcp", sQPAddr)
	if err != nil {
		log.Fatal(err)
	}
	capconn := rpc.NewConn(rpc.NewStreamTransport(conn), nil)
	client := crossXDT.StreamData(capconn.Bootstrap(context.Background()))
	f, _ := client.ServeBroadcastData(ctx, func(ps crossXDT.StreamData_serveBroadcastData_Params) error {
		ps.SetKey(key)
		return nil
	})
	//defer release()

	res, err := f.Struct()
	if err != nil {
		log.Fatal(err)
	}
	payload, err := res.Payload()
	if err != nil {
		log.Fatal(err)
	}

	log.Info("DST: ", payload[0:9], payload[len(payload)-9:])
	return payload, nil
}

// StartDstServer starts DstQP server
func StartDstServer(config utils.Config, handler func([]byte) ([]byte, bool)) {

	lis, err := net.Listen("tcp", config.DstServerPort)
	if err != nil {
		log.Fatalf("DST: failed to listen: %v", err)
	}
	var server *grpc.Server
	if config.TracingEnabled {
		server = grpc.NewServer(grpc.UnaryInterceptor(otelgrpc.UnaryServerInterceptor()),
			grpc.StreamInterceptor(otelgrpc.StreamServerInterceptor()))
	} else {
		server = grpc.NewServer()
	}

	conn, err := net.Dial("tcp", config.DQPServerHostname+config.DQPServerPort)
	if err != nil {
		log.Fatal(err)
	}
	capconn := rpc.NewConn(rpc.NewStreamTransport(conn), nil)

	s := downXDTServer{}
	s.config = config
	s.handler = handler
	s.conn = capconn
	downXDT.RegisterXDTtoFnServer(server, s)

	log.Infoln("DST: start server")
	if err := server.Serve(lis); err != nil {
		log.Fatalf("DST: failed to serve: %v", err)
	}
}

File: sdk/python/destination.py
# MIT License

# Copyright (c) 2021 Shyam Jesalpura and EASE lab

# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:

# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.

# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

import sys
import os
from datetime import datetime
from socket import socket

# adding gRPC sources to the system path
sys.path.insert(0, os.getcwd() + '/../../proto/downXDT')
sys.path.insert(0, os.getcwd() + '/../../proto/crossXDT')

from concurrent import futures
import grpc
import logging as log
import downXDT_pb2_grpc
import downXDT_pb2
import crossXDT_pb2_grpc
import crossXDT_pb2
from utils import loadConfig, STORE_FORWARD
import capnp
crossXDT_capnp = capnp.load('crossXDT.py.capnp')
capnp.remove_event_loop()
capnp.create_event_loop(threaded=True)


# XDTtoFnServicer is to be called by dQP to invoke DstFn
class XDTtoFnServicer(downXDT_pb2_grpc.XDTtoFnServicer):
    def __init__(self, dstHandler, payloadFetcher):
        self.dstHandler = dstHandler
        self.payloadFetcher = payloadFetcher

    def XDTFnCall(self, request, context):
        log.info("DST: received invocation call %s", request.XDTJSON)
        metadict = dict(context.invocation_metadata())
        if metadict['is_xdt'] == "true":
            key = metadict['key']
            # fetch data from dQP
            start = datetime.now()
            log.info("Fetching payload using key %s", key)
            payloadBytes = self.payloadFetcher.FetchFromDQP(key)
            duration = datetime.now() - start
            log.info("XDT pull keys took %f seconds", duration.total_seconds())
            # call destination function
            message, ok = self.dstHandler(payloadBytes)
            return downXDT_pb2.InvocationResponse(message=message, ok=ok)
        else:
            return downXDT_pb2.InvocationResponse(message=b'', ok=False)


class Fetcher:
    def __init__(self, config):
        self.config = config

# FetchFromDQP fetches data from dQP to DstFn
    def FetchFromDQP(self, key):
        # serverAddr = self.config['DQPServerHostname']+self.config['DQPServerPort']
        log.info("[dst] making a call to dqp @ %s using key %s", self.config['DQPServerHostname']+self.config['DQPServerPort'], key)
        # client = capnp.TwoPartyClient(self.config['DQPServerHostname']+self.config['DQPServerPort'])
        with socket() as s:
            s.connect((self.config['DQPServerHostname'].encode(), int(self.config['DQPServerPort'][1:])))
            client = capnp.TwoPartyClient(s)
            packet = client.bootstrap().cast_as(crossXDT_capnp.StreamData)

            request = packet.serveData_request()
            request.key = key

            # Send it, which returns a promise for the result (without blocking).
            get_promise = request.send()
            log.info("[dst] waiting for a response from dqp")
            response = get_promise.wait()
            return response.payload


# Get fetches data from src
def Get(capability, config):
    log.info("fetching payload using capability %s", capability)
    key = capability
    splitString = capability.split("|", 1)
    srcAddr = splitString[1]
    metadata = (
        ('is_xdt', 'true'),
        ('key', key),
        ('src_addr', srcAddr),
        ('routing', STORE_FORWARD),
    )
    client = capnp.TwoPartyClient(srcAddr)
    packet = client.bootstrap().cast_as(crossXDT_capnp.StreamData)
    log.debug("Getting from server... ")

    request = packet.serveData_request()
    request.key = key

    # Send it, which returns a promise for the result (without blocking).
    get_promise = request.send()
    response = get_promise.wait()
    return response.payload


# Get fetches data from src
def BroadcastGet(capability, config):
    log.info("fetching payload using capability %s", capability)
    key = capability
    splitString = capability.split("|", 1)
    srcAddr = splitString[1]
    metadata = (
        ('is_xdt', 'true'),
        ('key', key),
        ('src_addr', srcAddr),
        ('routing', STORE_FORWARD),
    )
    client = capnp.TwoPartyClient(srcAddr)
    packet = client.bootstrap().cast_as(crossXDT_capnp.StreamData)
    log.debug("Getting from server... ")

    request = packet.serveBroadcastData_request()
    request.key = key

    # Send it, which returns a promise for the result (without blocking).
    get_promise = request.send()
    response = get_promise.wait()
    log.info("fetched payload of size %s", len(response.payload))
    return response.payload


# StartDstServer starts DstQP server
def StartDstServer(config, dstHandler):

    payloadFetcher = Fetcher(config=config)
    server = grpc.server(futures.ThreadPoolExecutor())
    xdtServicer = XDTtoFnServicer(dstHandler=dstHandler, payloadFetcher=payloadFetcher)
    downXDT_pb2_grpc.add_XDTtoFnServicer_to_server(
        xdtServicer, server)
    server.add_insecure_port("[::]"+config['DstServerPort'])
    server.start()

    server.wait_for_termination()


if __name__ == '__main__':
    log.basicConfig(level=log.INFO)

    def handler(payload):
        log.info("destination received payload of length %d", len(payload))
        return b"sample response", True

    StartDstServer(loadConfig(), handler)
